Задание 2.4

В аптеку поступают партии лекарств. Их названия находятся в списке names, количество единиц товара находится в списке counts.

Например:

names = ['chlorhexidine', 'cyntomycin', 'afobazol'] counts = [15, 18, 7]

Напишите функцию create_medications(names, counts), которая создает Series medications, индексами которой являются названия лекарств names, а значениями - их количество в поставке counts.

А также напишите функцию get_percent(medications, name), которая возвращает долю количества товара с именем name от общего количества товаров в поставке в процентах.

Примечание. Не забудьте ипортировать библиотеки.

Пример:


names=['chlorhexidine', 'cyntomycin', 'afobazol']
counts=[15, 18, 7]
medications = create_medications(names, counts)

print(get_percent(medications, "chlorhexidine")) #37.5

In [55]:
import pandas as pd

def create_medications(names, counts):
    """
    Создает Series, где названия лекарств - это индексы, а их количество - значения.
    """
    return pd.Series(counts, index=names)

def get_percent(medications, name):
    """
    Возвращает процентное соотношение количества указанного лекарства
    к общему количеству всех лекарств в поставке.
    """
    if name not in medications:
        return 0  # Если лекарства нет в списке, возвращаем 0%
    total = medications.sum()
    return (medications[name] / total) * 100

# Пример использования:
names = ['chlorhexidine', 'cyntomycin', 'afobazol']
counts = [15, 18, 7]
medications = create_medications(names, counts)

print(get_percent(medications, "chlorhexidine"))  # 37.5


37.5


Задание 3.5 

Вы работаете аналитиком в компании ScienceYou. Ваша задача — проанализировать чистую прибыль.

Доходы (income), расходы (expenses) и годы (years), соответствующие им, предоставлены вам в виде списков.

Например:


income = [478, 512, 196]
expenses = [156, 130, 270]
years = [2018, 2019, 2020]
Создайте функцию create_companyDF(income, expenses, years), которая возвращает DataFrame, составленный из входных данных со столбцами Income и Expenses и индексами, соответствующими годам рассматриваемого периода.

Пример такого DataFrame представлен ниже.


    Income  Expenses
2018    478     156
2019    512     130
2020    196     270
Также напишите функцию get_profit(df, year), которая возвращает разницу между доходом и расходом, записанными в таблице df, за год year.

Примечание. Если информация за запрашиваемый год не указана в вашей таблице, вам необходимо вернуть None.

In [56]:
import pandas as pd

def create_companyDF(income, expenses, years):
    data = {'Income': income, 'Expenses': expenses}
    df = pd.DataFrame(data, index=years)
    return df

def get_profit(df, year):
    if year in df.index:
        return df.loc[year, 'Income'] - df.loc[year, 'Expenses']
    return None

# Пример использования
income = [478, 512, 196]
expenses = [156, 130, 270]
years = [2018, 2019, 2020]

df = create_companyDF(income, expenses, years)
print(df)

profit_2019 = get_profit(df, 2019)
print(f"Разница между доходом и расходом за 2019 год: {profit_2019}")


      Income  Expenses
2018     478       156
2019     512       130
2020     196       270
Разница между доходом и расходом за 2019 год: 382


Данные, с которыми мы будем работать — это обработанный датасет об объектах недвижимости в Мельбурне (Австралия) и его пригородах.

Данные представляют собой таблицу, в которой содержится 23 столбца:

index — номер строки
Suburb — наименование пригорода
Address — адрес
Rooms — количество комнат в помещении
Type — тип здания (h — дом, коттедж, вилла, терраса; u — блочный, дуплексный дом; t — таунхаус)
Price — цена помещения
Method — метод продажи 
SellerG — риэлторская компания
Date — дата продажи (в формате день/месяц/год)
Distance — расстояния до объекта от центра Мельбурна 
Postcode — почтовый индекс
Bedroom — количество спален
Bathroom — количество ванных комнат
Car — количество парковочных мест
Landsize — площадь прилегающей территории
BuildingArea — площадь здания
YearBuilt — год постройки
CouncilArea — региональное управление
Lattitude — географическая широта
Longitude — географическая долгота
Regionname — наименование района Мельбурна
Propertycount — количество объектов недвижимости в районе, выставленных на продажу
Coordinates — широта и долгота, объединённые в кортеж

In [57]:
#Прочитаем наши данные о недвижимости из csv-файла и запишем результирующий DataFrame в переменную melb_data:

melb_data = pd.read_csv('data/melb_data.csv', sep=',')
display(melb_data)

index         Suburb           Address  Rooms Type      Price Method  \
0          0     Abbotsford      85 Turner St      2    h  1480000.0      S   
1          1     Abbotsford   25 Bloomburg St      2    h  1035000.0      S   
2          2     Abbotsford      5 Charles St      3    h  1465000.0     SP   
3          3     Abbotsford  40 Federation La      3    h   850000.0     PI   
4          4     Abbotsford       55a Park St      4    h  1600000.0     VB   
...      ...            ...               ...    ...  ...        ...    ...   
13575  13575  Wheelers Hill      12 Strada Cr      4    h  1245000.0      S   
13576  13576   Williamstown     77 Merrett Dr      3    h  1031000.0     SP   
13577  13577   Williamstown       83 Power St      3    h  1170000.0      S   
13578  13578   Williamstown      96 Verdon St      4    h  2500000.0     PI   
13579  13579     Yarraville        6 Agnes St      4    h  1285000.0     SP   

        SellerG        Date  Distance  ...  Car  Landsize  BuildingArea  \
0        Biggin   3/12/2016       2.5  ...  1.0     202.0         126.0   
1        Biggin   4/02/2016       2.5  ...  0.0     156.0          79.0   
2        Biggin   4/03/2017       2.5  ...  0.0     134.0         150.0   
3        Biggin   4/03/2017       2.5  ...  1.0      94.0         126.0   
4        Nelson   4/06/2016       2.5  ...  2.0     120.0         142.0   
...         ...         ...       ...  ...  ...       ...           ...   
13575     Barry  26/08/2017      16.7  ...  2.0     652.0         126.0   
13576  Williams  26/08/2017       6.8  ...  2.0     333.0         133.0   
13577     Raine  26/08/2017       6.8  ...  4.0     436.0         126.0   
13578   Sweeney  26/08/2017       6.8  ...  5.0     866.0         157.0   
13579   Village  26/08/2017       6.3  ...  1.0     362.0         112.0   

       YearBuilt  CouncilArea  Lattitude  Longtitude  \
0         1970.0        Yarra  -37.79960   144.99840   
1         1900.0        Yarra  -37.80790   144.99340   
2         1900.0        Yarra  -37.80930   144.99440   
3         1970.0        Yarra  -37.79690   144.99690   
4         2014.0        Yarra  -37.80720   144.99410   
...          ...          ...        ...         ...   
13575     1981.0          NaN  -37.90562   145.16761   
13576     1995.0          NaN  -37.85927   144.87904   
13577     1997.0          NaN  -37.85274   144.88738   
13578     1920.0          NaN  -37.85908   144.89299   
13579     1920.0          NaN  -37.81188   144.88449   

                       Regionname  Propertycount           Coordinates  
0           Northern Metropolitan         4019.0    -37.7996, 144.9984  
1           Northern Metropolitan         4019.0    -37.8079, 144.9934  
2           Northern Metropolitan         4019.0    -37.8093, 144.9944  
3           Northern Metropolitan         4019.0    -37.7969, 144.9969  
4           Northern Metropolitan         4019.0    -37.8072, 144.9941  
...                           ...            ...                   ...  
13575  South-Eastern Metropolitan         7392.0  -37.90562, 145.16761  
13576        Western Metropolitan         6380.0  -37.85927, 144.87904  
13577        Western Metropolitan         6380.0  -37.85274, 144.88738  
13578        Western Metropolitan         6380.0  -37.85908, 144.89299  
13579        Western Metropolitan         6543.0  -37.81188, 144.88449  

[13580 rows x 23 columns]

Задание 5.1

In [58]:
#Какова цена объекта недвижимости под индексом 15?
melb_data = pd.read_csv('data/melb_data.csv', sep=',')
price_index_15 = melb_data.loc[15, 'Price']
display(price_index_15)

1310000.0

In [59]:
#Когда был продан объект под индексом 90?
melb_data = pd.read_csv('data/melb_data.csv', sep=',')
price_index_90 = melb_data.loc[90, 'Date']
display(price_index_90)

'10/09/2016'

Задание 5.3

Во сколько раз площадь прилегающей территории, на которой находится здание
с индексом 3521, больше площади участка, на котором находится здание 
с индексом 1690? Ответ округлите до целого числа.

In [ ]:
melb_data = pd.read_csv('data/melb_data.csv', sep=',')
# Извлекаем площадь прилегающей территории (Landsize) для объектов 
# с индексами 3521 и 1690.
landsize_3521 = melb_data.loc[3521, 'Landsize']
landsize_1690 = melb_data.loc[1690, 'Landsize']
# Делим одно значение на другое.
# Округляем результат до целого числа с помощью round().
ratio = round(landsize_3521 / landsize_1690)
display(ratio)

3

Задание 6.1

Дана таблица data = pd.DataFrame([[0,1], [1, 0], [1, 1]], columns=['А', 'B']). Что выведет код print(data.shape)?

Чтобы узнать размер таблицы — количество строк и количество столбцов. Это можно сделать с помощью атрибута shape, который возвращает кортеж с количеством строк и столбцов:

In [64]:
data = pd.DataFrame([[0,1], [1, 0], [1, 1]], columns=['А', 'B'])
print(data.shape)


(3, 2)


Задание 6.2

Какой из нижеперечисленных вариантов кода позволяет получить последние 15 строк DataFrame melb_data?

melb_data.tail(15)

при этом head() верхние (по умолчанию 5)
tail() нижние пять 

Задание 6.3
melb_data.info()
Какой тип данных имеет столбец Coordinates?

 22  Coordinates    13580 non-null  object 
посмотрели с помощью метода info()
Данный метод выводит:

информацию об индексах;
информацию об общем количестве столбцов;
таблицу, в которой содержится информация об именах столбцов (Column), количестве непустых значений (Non-Null Count) в каждом столбце и типе данных столбца (Dtype), количестве столбцов, в которых используется определённый тип данных;
количество оперативной памяти в мегабайтах, которое тратится на хранение данных.

Задание 6.4

Какое утверждение о типе данных object верно?

Это тип данных, которым в Pandas кодируются текстовые или другие нечисловые значения (кортежи, списки и т. д.)

Задание 6.5
melb_data.info()
Сколько пропущенных значений в столбце CouncilArea?
ответ 1369

 16  YearBuilt      13580 non-null  int64  
 17  CouncilArea    12211 non-null  object 
 18  Lattitude      13580 non-null  float64

 13580-12211= 1369

Задание 6.6
melb_data.info()
Сколько столбцов после преобразования типов имеет целочисленный тип данных?



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          13580 non-null  int64  
 1   Suburb         13580 non-null  object 
 2   Address        13580 non-null  object 
 3   Rooms          13580 non-null  int64  
 4   Type           13580 non-null  object 
 5   Price          13580 non-null  float64
 6   Method         13580 non-null  object 
 7   SellerG        13580 non-null  object 
 8   Date           13580 non-null  object 
 9   Distance       13580 non-null  float64
 10  Postcode       13580 non-null  int64  
 11  Bedroom        13580 non-null  int64  
 12  Bathroom       13580 non-null  int64  
 13  Car            13580 non-null  int64  
 14  Landsize       13580 non-null  float64
 15  BuildingArea   13580 non-null  float64
 16  YearBuilt      13580 non-null  int64  
 17  CouncilArea    12211 non-null  object 
 18  Lattitude      13580 non-null  float64
 19  Longtitude     13580 non-null  float64
...
 21  Propertycount  13580 non-null  int64  
 22  Coordinates    13580 non-null  object 
dtypes: float64(6), int64(8), object(9)
memory usage: 2.4+ MB

Задание 6.7

Выберите верные утверждения в отношении таблицы melb_data:


Информацию о статистических параметрах столбцов с данными типа object можно получить с помощью метода describe() с параметром include=['object'].


A Наименьшее число объектов расположено в районе Southern Metropolitan
B Мельбурн и его пригороды разделены как минимум на восемь районов
C У некоторых различных адресов в таблице совпадают географические координаты
D В пригороде Reservoir расположено наибольшее число проданных объектов - 359


ни одно утверждение не верно!!!

Задание 6.9

Сколько процентов от общего количества домов составляют таунхаусы (тип объекта — t)?
Ответ округлите до целого числа. Не указывайте знак %.

In [65]:
# Посчитаем общее количество записей в melb_data
total_count = len(melb_data)  # Общее количество объектов
# Вычисляем количество таунхаусов (Type == 't').
townhouse_count = (melb_data['Type'] == 't').sum()  # Количество таунхаусов
# Найти процентное соотношение и округлить до целого числа.
percentage = round((townhouse_count / total_count) * 100)
print(percentage)

8
